File: tests/clusters/gke.go
// Copyright 2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package clusters

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	container "cloud.google.com/go/container/apiv1"
	"cloud.google.com/go/container/apiv1/containerpb"
	"github.com/sethvargo/go-githubactions"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
	"google.golang.org/api/impersonate"
	"google.golang.org/api/option"

	"k8s.io/client-go/tools/clientcmd"
	"k8s.io/client-go/tools/clientcmd/api"
)

type GKECluster struct {
	name            string
	previousContext string
	client          *container.ClusterManagerClient
	kubeconfig      api.Config
}

var _ Cluster = GKECluster{}

func NewGKECluster(namePrefix string, kubeconfig api.Config) (Cluster, error) {
	// It normally takes 7-10 minutes to create a cluster.
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Minute)
	defer cancel()

	// Login to GCP.

	clusterName := namePrefix + randString()
	projectID := "pulumi-k8s-provider"
	location := "us-west1-a"

	serviceAccount := "pulumi-ci@pulumi-k8s-provider.iam.gserviceaccount.com"
	audience := fmt.Sprintf("https://iam.googleapis.com/projects/%s/locations/global/workloadIdentityPools/pulumi-ci/pulumi-ci", 637339343727) // Project number?

	// Default to local credentials.
	auth, err := google.DefaultTokenSource(ctx)
	if err != nil {
		return nil, fmt.Errorf("getting default token source: %w", err)
	}
	if os.Getenv("ACTIONS_ID_TOKEN_REQUEST_TOKEN") != "" {
		// If we're running as a GitHub action, grab an ID token so we can use
		// that as our token source.
		token, err := githubactions.GetIDToken(ctx, audience)
		if err != nil {
			return nil, fmt.Errorf("getting ID token: %w", err)
		}
		auth = oauth2.StaticTokenSource(&oauth2.Token{
			AccessToken: token,
			TokenType:   "urn:ietf:params:oauth:token-type:jwt",
		})
	}

	token, err := auth.Token()
	if err != nil {
		return nil, fmt.Errorf("getting token: %w", err)
	}

	ts, err := impersonate.CredentialsTokenSource(ctx,
		impersonate.CredentialsConfig{
			Scopes:          []string{"https://www.googleapis.com/auth/cloud-platform"},
			TargetPrincipal: serviceAccount,
			Subject:         token.AccessToken,
		},
		option.WithTokenSource(auth),
		option.WithAudiences(audience),
	)
	if err != nil {
		return nil, fmt.Errorf("getting impersonate source: %w", err)
	}

	// Create the GKE client
	client, err := container.NewClusterManagerClient(ctx, option.WithTokenSource(ts))
	if err != nil {
		log.Fatalf("creating GKE client: %v", err)
	}
	defer client.Close()

	// Define the cluster
	cluster := &containerpb.Cluster{
		Name:             clusterName,
		InitialNodeCount: 1,
		NodeConfig: &containerpb.NodeConfig{
			MachineType: "n1-standard-2",
		},
		Location: location,
	}

	// Create the cluster
	createClusterReq := &containerpb.CreateClusterRequest{
		Parent:  fmt.Sprintf("projects/%s/locations/%s", projectID, location),
		Cluster: cluster,
	}

	fmt.Println("creating GKE cluster...")
	_, err = client.CreateCluster(ctx, createClusterReq)
	if err != nil {
		return nil, fmt.Errorf("creating GKE cluster: %w", err)
	}

	fmt.Println("waiting for cluster to become ready..")

	for {
		req := &containerpb.GetClusterRequest{
			Name: fmt.Sprintf("projects/%s/locations/%s/clusters/%s", projectID, location, clusterName),
		}

		cluster, err = client.GetCluster(ctx, req)
		if err != nil {
			return nil, fmt.Errorf("failed to get cluster details: %w", err)
		}

		fmt.Println("cluster is", cluster.Status.String())

		if cluster.Status == containerpb.Cluster_RUNNING {
			break
		}

		select {
		case <-time.After(10 * time.Second):
			continue
		case <-ctx.Done():
			return nil, ctx.Err()
		}
	}

	// Add the cluster to kubeconfig.

	token, err = ts.Token()
	if err != nil {
		return nil, fmt.Errorf("getting token for kubeconfig: %w", err)
	}

	if kubeconfig.Clusters == nil {
		kubeconfig.Clusters = map[string]*api.Cluster{}
	}
	kubeconfig.Clusters[clusterName] = &api.Cluster{
		Server:                   fmt.Sprintf("https://%s", cluster.Endpoint),
		CertificateAuthorityData: []byte(cluster.MasterAuth.ClusterCaCertificate),
	}
	if kubeconfig.Contexts == nil {
		kubeconfig.Contexts = map[string]*api.Context{}
	}
	kubeconfig.Contexts[clusterName] = &api.Context{
		Cluster:  clusterName,
		AuthInfo: clusterName,
	}
	if kubeconfig.AuthInfos == nil {
		kubeconfig.AuthInfos = map[string]*api.AuthInfo{}
	}
	kubeconfig.AuthInfos[clusterName] = &api.AuthInfo{
		Token: token.AccessToken, // Probably won't work...
		/*
			Exec: &api.ExecConfig{
				Command: "gke-gcloud-auth-plugin", // Not installed...
			},
		*/
	}

	previousContext := kubeconfig.CurrentContext
	kubeconfig.CurrentContext = clusterName

	if err := clientcmd.WriteToFile(kubeconfig, "~/.kube/config"); err != nil {
		return nil, fmt.Errorf("writing kubeconfig: %w", err)
	}

	return GKECluster{name: clusterName, previousContext: previousContext, client: client}, nil
}

func (c GKECluster) Name() string {
	return c.name
}

func (c GKECluster) Delete() error {
	_, err := c.client.DeleteCluster(context.Background(), &containerpb.DeleteClusterRequest{
		Name: c.name,
	})
	if err != nil {
		return fmt.Errorf("deleting cluster: %w", err)
	}

	// Remove the cluster's information and restore previous context.
	c.kubeconfig.CurrentContext = c.previousContext
	delete(c.kubeconfig.Clusters, c.name)
	delete(c.kubeconfig.AuthInfos, c.name)
	delete(c.kubeconfig.Contexts, c.name)

	if err := clientcmd.WriteToFile(c.kubeconfig, "~/.kube/config"); err != nil {
		return fmt.Errorf("writing kubeconfig: %w", err)
	}

	return nil
}

File: tests/ci-cluster/gke.ts
// Copyright 2016-2022, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as gcp from "@pulumi/gcp";
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";
import * as config from "./config";

export class GkeCluster extends pulumi.ComponentResource {
    public cluster: gcp.container.Cluster;
    public kubeconfig: pulumi.Output<string>;
    public provider: k8s.Provider;

    constructor(name: string,
                opts: pulumi.ComponentResourceOptions = {}) {
        super("pulumi-kubernetes:ci:GkeCluster", name, {}, opts);

        // Use the latest 1.32.x engine version.
        const engineVersion = "1.32";

        // Create the GKE cluster.
        const k8sCluster = new gcp.container.Cluster("ephemeral-ci-cluster", {
            initialNodeCount: config.nodeCount,
            nodeVersion: engineVersion,
            minMasterVersion: engineVersion,
            nodeConfig: {
                machineType: config.nodeMachineType,
                oauthScopes: [
                    "https://www.googleapis.com/auth/compute",
                    "https://www.googleapis.com/auth/devstorage.read_only",
                    "https://www.googleapis.com/auth/logging.write",
                    "https://www.googleapis.com/auth/monitoring"
                ],
            },
            project: config.gcpProject,
            location: config.gcpLocation,
            // Enable network policy addon to test network policy resources.
            addonsConfig: {
              networkPolicyConfig: {
                disabled: false,
              },
            },
            networkPolicy: {
              enabled: true,
              provider: "CALICO",
            },
        }, {parent: this});
        this.cluster = k8sCluster;

        // Manufacture a GKE-style Kubeconfig. Note that this is slightly "different" because of the way GKE requires
        // gcloud to be in the picture for cluster authentication (rather than using the client cert/key directly).
        this.kubeconfig = pulumi.all([k8sCluster.name, k8sCluster.endpoint, k8sCluster.masterAuth]).apply(
            ([name, endpoint, auth]) => {
                const context = `${config.gcpProject}_${config.gcpZone}_${name}`;
                return `apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ${auth.clusterCaCertificate}
    server: https://${endpoint}
  name: ${context}
contexts:
- context:
    cluster: ${context}
    user: ${context}
  name: ${context}
current-context: ${context}
kind: Config
preferences: {}
users:
- name: ${context}
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      command: gke-gcloud-auth-plugin
      installHint: Install gke-gcloud-auth-plugin for use with kubectl by following
        https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke
      provideClusterInfo: true`;
            });

        // Export a Kubernetes provider instance that uses our cluster from above.
        this.provider = new k8s.Provider("gke", {kubeconfig: this.kubeconfig}, {parent: this});
    }
}

